mito-ai/mito_ai/gemini_client.py

```diff
@@ -17,71 +88,80 @@ def __init__(self, api_key: Optional[str], model: str):
             self.client = genai.Client(api_key=api_key)
 
     async def request_completions(
-        self, 
-        messages: List[Dict[str, Any]], 
+        self,
+        messages: List[Dict[str, Any]],
         response_format_info: Optional[ResponseFormatInfo] = None,
         message_type: MessageType = MessageType.CHAT
     ) -> str:
         try:
-            contents = self.convert_messages_for_gemini(messages)
             config = None
+
+            # Extract system instructions and Gemini-compatible contents
+            system_instructions, contents = extract_system_instruction_and_contents(messages)
+
             if response_format_info:
                 config = {
                     "response_mime_type": "application/json",
                     "response_schema": AgentResponse.model_json_schema()
                 }
-            # Build provider_data once
+
             provider_data = get_gemini_completion_function_params(
                 model=self.model if not response_format_info else INLINE_COMPLETION_MODEL,
                 contents=contents,
                 message_type=message_type,
                 config=config,
                 response_format_info=response_format_info
             )
+
             if self.api_key:
-                # Unpack provider_data for direct API call
-                response = self.client.models.generate_content(
-                    model=provider_data["model"],
-                    contents=provider_data["contents"]
-                )
+                if provider_data.get("config"):
+                    response = self.client.models.generate_content(
+                        model=provider_data["model"],
+                        contents=contents,
+                        config=GenerateContentConfig(
+                            system_instruction=system_instructions,
+                            response_mime_type=provider_data["config"].get("response_mime_type") if provider_data.get("config") else None,
+                            response_schema=provider_data["config"].get("response_schema") if provider_data.get("config") else None
+                        )
+                    )
+                else:
+                    response = self.client.models.generate_content(
+                        model=provider_data["model"],
+                        contents=contents,
+                        config=GenerateContentConfig(
+                            system_instruction=""   # No system instruction needed for inline completions
+                        )
+                    )
```

Comment:
Would this simplify things?

<img width="767" alt="Screenshot 2025-06-10 at 4 08 00 PM" src="https://github.com/user-attachments/assets/70fc32d5-dceb-435f-806e-3b1e92314a91" />